packages/synchro-charts/src/components/sc-dial/sc-dial-base/sc-dial-base.tsx

```diff
@@ -0,0 +1,186 @@
+import { NO_VALUE_PRESENT } from '../../common/terms';
+import { DialLoading } from './sc-dial-loading';
+import { TextSizeConfig, sizeContent, StatusProgress, LengthToSize, SizeFont } from './util';
+
+const title = (dataStream: { detailedName?: any; name?: any } | null | false) => {
+  if (dataStream) {
+    return dataStream.detailedName || dataStream.name;
+  }
+  return null;
+};
+
+const UNITMINLENGTH = 1;
+@Component({
+  tag: 'sc-dial-base',
+  styleUrl: 'sc-dial-base.css',
+  shadow: false,
+})
+export class ScDialBase {
+  @Prop() viewport: ViewPortConfig;
+  @Prop() breachedThreshold?: Threshold;
+  @Prop() valueColor?: string; // css color string
+
+  @Prop() alarmStream?: DataStream;
+
+  @Prop() propertyStream?: DataStream;
+  @Prop() propertyPoint?: DataPoint<Primitive>;
+  @Prop() size?: MinimalSizeConfig = { width: 0, height: 0 };
+
+  @Prop() isLoading?: boolean = false;
+
+  @State() fontSize: TextSizeConfig = sizeContent.XL;
+
+  getColorByPercent = (percent: number): string => {
+    const LEVEL1 = 0.3333;
+    const LEVEL2 = 0.6667;
+    let color = '';
+    if (percent < LEVEL1) {
+      color = StatusProgress.CRITICAL;
+    } else if (percent < LEVEL2) {
+      color = StatusProgress.WARNING;
+    } else {
+      color = StatusProgress.NORMAL;
+    }
+    return color;
+  };
+
+  getFontSize = (length: number): string => {
+    return LengthToSize[length] || SizeFont.XS;
+  };
+
+  @Watch('propertyStream')
+  getSize(newVal: DataStream) {
+    const dataLength = newVal.data.length;
+    const unitLength = newVal.unit?.length || UNITMINLENGTH;
+    const valueLength = dataLength && newVal.data[dataLength - 1].y.toString().length;
+    const textLength = valueLength + unitLength;
+
+    this.fontSize = sizeContent[this.getFontSize(textLength)];
+  }
+
+  render() {
+    const icon = this.breachedThreshold ? this.breachedThreshold.icon : undefined;
+    const { yMin = 0, yMax = 0 } = this.viewport;
+    const propertyStream = this.propertyStream ? isNumberDataStream(this.propertyStream) && this.propertyStream : null;
+    const point = propertyStream ? this.propertyPoint : null;
+    const error = propertyStream ? propertyStream.error : 'Only numbers are supported';
+
+    const percent = point ? (point.y as number) / (yMax - yMin) : 0;
+    const labelColor =
+      this.valueColor || (this.breachedThreshold ? this.getColorByPercent(percent) : StatusProgress.BLUE);
+    const left = 760 * (1 - percent);
+    const right = 760 - left;
+    const stream = propertyStream;
+    const unit = propertyStream && propertyStream.unit;
+
+    return (
+      <sc-dial-tooltip
+        title={title(propertyStream)}
+        propertyPoint={this.propertyPoint}
+        alarmPoint={this.alarmStream && this.propertyPoint}
+        breachedThreshold={this.breachedThreshold}
+        unit={unit || '%'}
+        value={unit ? (point?.y as number) : percent * 100}
+        color={labelColor}
+      >
+        <div class="sc-dialbase-container">
```

Comment:
is there some way we could break this into sub-components to improve readability and maintainability?

Reply:
I add a component `DialSVG`.